### ISSUE TITLE
How many label data were used in the train workflow?

### ISSUE THREAD

**Issue by yaohuatj** [shaohua0116/SSGAN-Tensorflow#3]
I mean label data used in the model, do not used testing the accuracy!

**Comment by wangguanan** [comment]
I have the same question. when I read the code, I found author use all labeled data to train the both s_loss and gan loss. Labeled data are used to optimize s_loss, at the same time the same data are also used to train d_loss and g_loss(amazingly course label information is also used). It means that all data are used in a supervised way.  It is a absolutely supervised model, never semi supervised model. 
Are you kidding me? I have read the code for day and want to modify it to adapt to my model.

Parameter [label] is supervised information, and every loss(s_loss, d_loss_real, d_loss_fake) used parameter label, so the model is absolutely not a semi supervised model. 
Code are shown in the figure below.

![image](https://user-images.githubusercontent.com/18496599/28236258-f590cef0-6953-11e7-9b02-456e5d117bc6.png)

**Comment by shaohua0116** [comment]
Thank you for raising your concerns. Please be aware of the reason why the model is called semi-supervised GAN. The model learns to exploit the samples generated by the generator, which have absolutely no label, to boost the performance of the task by improving generalization.
Please refer to the related papers such as [Improved Techniques for Training GANs](https://arxiv.org/abs/1606.03498) or [Good Semi-supervised Learning that Requires a Bad GAN](https://arxiv.org/abs/1705.09783) for further information. If you still get confused about the concept or the implementation, do let me know and I will try to help.

**Comment by wangguanan** [comment]
Thank for your quick answer. I hold the opposite view.
I think you misunderstand the conception 'semi - supervised'. 
OK, back to the paper < Improved Techniques for Training GANs >, if the paper's author also hold the view that semi-supervision is just the data from generator,  why he/she compared results with different numbers of labeled samples(such as cifar10 1k/2k/3k/5k but never 50k because 50k is all data in training set)?


![image](https://user-images.githubusercontent.com/18496599/28236308-797686f0-6955-11e7-8c79-bf927ab220b1.png)

**Comment by shaohua0116** [comment]
The reason is that the performance boost is believed to be more observable when the data is less sufficient.

**Comment by wangguanan** [comment]
Thank for your explanation, but i still insist on my viewpoint.
In CIFAR10 setup, your implementation use all 50k imagess and all their labels in training set to train your model?
and how about the performance of your implementation with 50k images but part of their labels(such as 4k) , have you estimate that setup? if not, can you make a prediction because your implementation get a very high accuracy( peak to 85% and converge to more than 80%+ compared to less than 80% in the paper)?

**Comment by shaohua0116** [comment]
Please note that changing the number of training data can be done with modifying the codes in ```./datasets```.

**Comment by wangguanan** [comment]
OK，I will training with a small dataset😬(•‿•)
One more question， isn't BN layer usually added before activation layer？ you do it oppositely but model still work，its your mistake or something else？

**Comment by shaohua0116** [comment]
The original paper proposing batch normalization prescribes using BN before activations. However, in practice, the opposite is usually true. Here's my explanation. The whole point of using BN is to normal the distribution of features. Some of these features coming out of convolutional layers might be negative which is supposed to be truncated by activations such as ReLU. Therefore, adding BN before activations can be biased towards these negative values. On the other hand, adding BN after activations normalizes the distribution features without biasing them with features that are filtered out by the activation functions. Obviously, the latter makes more sense to me.

**Comment by wangguanan** [comment]
got it! All road lead to Rome!
Very glad to discuss with you， and I harvested a lot . Thank for your patience.(•‿•)(•‿•)

**Comment by howardgriffin** [comment]
大佬，666

**Comment by oduerr** [comment]
I am still not sure that this has been resolved. Has somebody managed to reproduced the numbers in Table 2 from above? In IMHO that has been done using all of the available images but just e.g. 4000 labels.

**Comment by wangguanan** [comment]
There must be some implementations with 4000 labels but not here, and with 4000 labels good performance still can be achieved.

**Comment by zhangqianhui** [comment]
@wangguanan Have you test it using parted-label data to train this model? for example, using 2-,50, 100 labeled example to train semi-gan, have you test it?

**Comment by wangguanan** [comment]
sorry, I do not test the code with part-labeled data. Actually, some paper do it, e.g. <UNSUPERVISED AND SEMI-SUPERVISED LEARNING WITH CATEGORICAL GENERATIVE ADVERSARIAL NETWORKS>, however no code of the method is access. you can implement it by yourself, I think it is not very difficult for  its setting is very detailed.

![image](https://user-images.githubusercontent.com/18496599/31043607-b7282c82-a5f1-11e7-8010-52fa07558e80.png)

**Comment by WayneDW** [comment]
> IMHO

What do you mean by IMHO?
